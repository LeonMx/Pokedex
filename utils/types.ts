export type NamedAPIResource = {
  name: string
  url: string
}

export type PokemonAbility = {
  is_hidden: boolean
  slot: number
  ability: NamedAPIResource
}

export type VersionGameIndex = {
  game_index: number
  version: NamedAPIResource
}

export type PokemonHeldItemVersion = {
  version: NamedAPIResource
  rarity: number
}

export type PokemonHeldItem = {
  item: NamedAPIResource
  version_details: PokemonHeldItemVersion[]
}

export type PokemonMoveVersion = {
  move_learn_method: NamedAPIResource
  version_group: NamedAPIResource
  level_learned_at: number
}

export type PokemonMove = {
  move: NamedAPIResource
  version_group_details: PokemonMoveVersion[]
}

export type PokemonSprites = {
  front_default: string
  front_shiny: string
  front_female: string
  front_shiny_female: string
  back_default: string
  back_shiny: string
  back_female: string
  back_shiny_female: string
}

export type PokemonStat = {
  stat: NamedAPIResource
  effort: number
  base_stat: number
}

export type PokemonType = {
  slot: number
  type: NamedAPIResource
}

export type Pokemon = {
  id: number
  name: string
  base_experience: number
  height: number
  is_default: boolean
  order: number
  weight: number
  abilities: PokemonAbility[]
  forms: NamedAPIResource[]
  game_indices: VersionGameIndex[]
  held_items: PokemonHeldItem[]
  location_area_encounters: string
  moves: PokemonMove[]
  sprites: PokemonSprites
  species: NamedAPIResource[]
  stats: PokemonStat[]
  types: PokemonType[]
}

export type TypeRelations = {
  no_damage_to: NamedAPIResource[]
  half_damage_to: NamedAPIResource[]
  double_damage_to: NamedAPIResource[]
  no_damage_from: NamedAPIResource[]
  half_damage_from: NamedAPIResource[]
  double_damage_from: NamedAPIResource[]
}

export type GenerationGameIndex = {
  game_index: number
  generation: NamedAPIResource
}

export type Name = {
  name: string
  language: NamedAPIResource
}

export type TypePokemon = {
  slot: number
  type: NamedAPIResource
}

export type Type = {
  id: number
  name: string
  damage_relations: TypeRelations
  game_indices: GenerationGameIndex[]
  generation: NamedAPIResource
  move_damage_class: NamedAPIResource
  names: Name[]
  pokemon: TypePokemon[]
  moves: NamedAPIResource[]
}

// this type represent pokemon data full(Pokemon) or partial(NamedAPIResource)
export type PokemonData = Pokemon | NamedAPIResource

// When pokemon param is Pokemon type, is concidered resolved
export const isPokemonResolved = (pokemon: PokemonData): pokemon is Pokemon => {
  return (pokemon as Pokemon).id !== undefined
}
