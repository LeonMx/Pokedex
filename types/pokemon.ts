export type NamedAPIResource = {
  name: string
  url: string
}

export type PokemonAbility = {
  is_hidden: boolean
  slot: number
  ability: NamedAPIResource
}

export type VersionGameIndex = {
  game_index: number
  version: NamedAPIResource
}

export type PokemonHeldItemVersion = {
  version: NamedAPIResource
  rarity: number
}

export type PokemonHeldItem = {
  item: NamedAPIResource
  version_details: Array<PokemonHeldItemVersion>
}

export type PokemonMoveVersion = {
  move_learn_method: NamedAPIResource
  version_group: NamedAPIResource
  level_learned_at: number
}

export type PokemonMove = {
  move: NamedAPIResource
  version_group_details: Array<PokemonMoveVersion>
}

export type PokemonSprites = {
  front_default: string
  front_shiny: string
  front_female: string
  front_shiny_female: string
  back_default: string
  back_shiny: string
  back_female: string
  back_shiny_female: string
}

export type PokemonStat = {
  stat: NamedAPIResource
  effort: number
  base_stat: number
}

export type PokemonType = {
  slot: number
  type: NamedAPIResource
}

export type Pokemon = {
  id: number
  name: string
  base_experience: number
  height: number
  is_default: boolean
  order: number
  weight: number
  abilities: Array<PokemonAbility>
  forms: Array<NamedAPIResource>
  game_indices: Array<VersionGameIndex>
  held_items: Array<PokemonHeldItem>
  location_area_encounters: string
  moves: Array<PokemonMove>
  sprites: PokemonSprites
  species: Array<NamedAPIResource>
  stats: Array<PokemonStat>
  types: Array<PokemonType>
}

// this type represent pokemon data full(Pokemon) or partial(NamedAPIResource)
export type PokemonData = Pokemon | NamedAPIResource

// When pokemon param is Pokemon type, is concidered resolved
export const isPokemonResolved = (pokemon: PokemonData): pokemon is Pokemon => {
  return (pokemon as Pokemon).id !== undefined
}
