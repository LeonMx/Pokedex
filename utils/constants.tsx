import { theme } from 'twin.macro'

export const POKEMON_TYPE = {
  NORMAL: 'normal',
  FIGHTING: 'fighting',
  FLYING: 'flying',
  POISON: 'poison',
  GROUND: 'ground',
  ROCK: 'rock',
  BUG: 'bug',
  GHOST: 'ghost',
  STEEL: 'steel',
  FIRE: 'fire',
  WATER: 'water',
  GRASS: 'grass',
  ELECTRIC: 'electric',
  PSYCHIC: 'psychic',
  ICE: 'ice',
  DRAGON: 'dragon',
  DARK: 'dark',
  FAIRY: 'fairy',
  UNKNOWN: 'unknown',
  SHADOW: 'shadow',
}

export const POKEMON_TYPE_COLOR = {
  [POKEMON_TYPE.NORMAL]: theme`colors.type-normal.DEFAULT`,
  [POKEMON_TYPE.FIGHTING]: theme`colors.type-fighting.DEFAULT`,
  [POKEMON_TYPE.FLYING]: theme`colors.type-flying.DEFAULT`,
  [POKEMON_TYPE.POISON]: theme`colors.type-poison.DEFAULT`,
  [POKEMON_TYPE.GROUND]: theme`colors.type-ground.DEFAULT`,
  [POKEMON_TYPE.ROCK]: theme`colors.type-rock.DEFAULT`,
  [POKEMON_TYPE.BUG]: theme`colors.type-bug.DEFAULT`,
  [POKEMON_TYPE.GHOST]: theme`colors.type-ghost.DEFAULT`,
  [POKEMON_TYPE.STEEL]: theme`colors.type-steel.DEFAULT`,
  [POKEMON_TYPE.FIRE]: theme`colors.type-fire.DEFAULT`,
  [POKEMON_TYPE.WATER]: theme`colors.type-water.DEFAULT`,
  [POKEMON_TYPE.GRASS]: theme`colors.type-grass.DEFAULT`,
  [POKEMON_TYPE.ELECTRIC]: theme`colors.type-electric.DEFAULT`,
  [POKEMON_TYPE.PSYCHIC]: theme`colors.type-psychic.DEFAULT`,
  [POKEMON_TYPE.ICE]: theme`colors.type-ice.DEFAULT`,
  [POKEMON_TYPE.DRAGON]: theme`colors.type-dragon.DEFAULT`,
  [POKEMON_TYPE.DARK]: theme`colors.type-dark.DEFAULT`,
  [POKEMON_TYPE.FAIRY]: theme`colors.type-fairy.DEFAULT`,
  [POKEMON_TYPE.UNKNOWN]: theme`colors.type-unknown.DEFAULT`,
  [POKEMON_TYPE.SHADOW]: theme`colors.type-shadow.DEFAULT`,
}

export const POKEMON_TYPE_COLOR_CONTRAST = {
  [POKEMON_TYPE.NORMAL]: theme`colors.type-normal.contrast`,
  [POKEMON_TYPE.FIGHTING]: theme`colors.type-fighting.contrast`,
  [POKEMON_TYPE.FLYING]: theme`colors.type-flying.contrast`,
  [POKEMON_TYPE.POISON]: theme`colors.type-poison.contrast`,
  [POKEMON_TYPE.GROUND]: theme`colors.type-ground.contrast`,
  [POKEMON_TYPE.ROCK]: theme`colors.type-rock.contrast`,
  [POKEMON_TYPE.BUG]: theme`colors.type-bug.contrast`,
  [POKEMON_TYPE.GHOST]: theme`colors.type-ghost.contrast`,
  [POKEMON_TYPE.STEEL]: theme`colors.type-steel.contrast`,
  [POKEMON_TYPE.FIRE]: theme`colors.type-fire.contrast`,
  [POKEMON_TYPE.WATER]: theme`colors.type-water.contrast`,
  [POKEMON_TYPE.GRASS]: theme`colors.type-grass.contrast`,
  [POKEMON_TYPE.ELECTRIC]: theme`colors.type-electric.contrast`,
  [POKEMON_TYPE.PSYCHIC]: theme`colors.type-psychic.contrast`,
  [POKEMON_TYPE.ICE]: theme`colors.type-ice.contrast`,
  [POKEMON_TYPE.DRAGON]: theme`colors.type-dragon.contrast`,
  [POKEMON_TYPE.DARK]: theme`colors.type-dark.contrast`,
  [POKEMON_TYPE.FAIRY]: theme`colors.type-fairy.contrast`,
  [POKEMON_TYPE.UNKNOWN]: theme`colors.type-unknown.contrast`,
  [POKEMON_TYPE.SHADOW]: theme`colors.type-shadow.contrast`,
}
