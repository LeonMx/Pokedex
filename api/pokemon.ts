import axios from 'axios'
import axiosRetry from 'axios-retry'
import { NamedAPIResource, Pokemon, Type } from 'utils/types'

const client = axios.create({ baseURL: 'https://pokeapi.co/api/v2' })
axiosRetry(client, { retries: 3 })

export type ResponseResourseList = {
  count: number
  next: string | null
  previous: string | null
  results: NamedAPIResource[]
}

type PageParams = {
  offset?: number
  limit?: number
}

type IdOrNameParam = string

const requestResourseLists = <T extends ResponseResourseList>(
  resourse: string
): ((params?: PageParams) => Promise<T>) => {
  return async (params) =>
    await client
      .get<T>(resourse, { params })
      .then(({ data }) => data)
}

const requestResourse = <T>(resourse: string): ((idOrName: IdOrNameParam) => Promise<T>) => {
  return async (idOrName) => await client.get<T>(`${resourse}/${idOrName}`).then(({ data }) => data)
}

export const getPokemons = requestResourseLists('pokemon')

export const getPokemon = requestResourse<Pokemon>('pokemon')

export const getTypes = requestResourseLists('type')

export const getType = requestResourse<Type>('type')
