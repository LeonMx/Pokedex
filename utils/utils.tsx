import { theme } from 'twin.macro'
import { STATS } from './constants'

export const mockConsoleMethod = (
  realConsoleMethod: CallableFunction
): ((message: string) => void) => {
  const ignoredMessages = ['test was not wrapped in act(...)']

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (message: string, ...args: any[]) => {
    const containsIgnoredMessage = ignoredMessages.some((ignoredMessage) =>
      message.includes(ignoredMessage)
    )

    if (!containsIgnoredMessage) {
      realConsoleMethod && realConsoleMethod(message, ...args)
    }
  }
}

export const calculateStatPoints = (
  name: string,
  baseStat: number
): {
  baseStat: number
  percentage: number
  minPoint: number
  maxPoint: number
} => {
  /* 
    For calculate max and min stat points:
  
    Formula for hp:
    max: BaseStat x 2 + 204
    min: BaseStat x 2 + 110

    Formula for other stats:
    max: ( BaseStat x 2 + 99 ) x 1.1
    min: ( BaseStat x 2 + 5 ) x 0.9

    https://bulbapedia.bulbagarden.net/wiki/Stat#Base_stats
  */
  const minPoint = name === STATS.HP ? baseStat * 2 + 110 : (baseStat * 2 + 5) * 0.9
  const maxPoint = name === STATS.HP ? baseStat * 2 + 204 : (baseStat * 2 + 99) * 1.1
  const percentage = (baseStat / 180) * 100
  return { baseStat, percentage, minPoint, maxPoint }
}

const statColor = {
  low: theme`colors.red.500`,
  neutral: theme`colors.yellow.300`,
  high: theme`colors.green.500`,
  highest: theme`colors.blue.500`,
}

export const getColorBaseStat = (baseStat: number): string => {
  if (baseStat < 50) return statColor.low

  if (baseStat < 100) return statColor.neutral

  if (baseStat < 150) return statColor.high

  return statColor.highest
}

export const getUnit = (num: number): string => (num * 0.1).toFixed(2)
