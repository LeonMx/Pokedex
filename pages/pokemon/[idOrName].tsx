import { FC } from 'react'
import { GetServerSideProps, NextPage } from 'next'
import Head from 'next/head'
import Image from 'next/image'
import tw, { styled, css } from 'twin.macro'
import Card from 'components/Card'
import TypeChip from 'components/TypeChip'
import { getPokemon } from 'api/pokemon'
import { Pokemon } from 'utils/types'
import DamageEffectiveness from 'components/DamageEffectiveness'
import ProgressBar from 'components/ProgressBar'
import { calculateStatPoints, getColorBaseStat, getUnit } from 'utils/utils'

export const TEST_ID_NAME_POKEMON = 'pokemon_name'
export const TEST_ID_NO_FOUND = 'no_found_pokemon'

type PokemonProps = { pokemon?: Pokemon }

const CardsContainer = styled.div<{ vertical?: boolean }>(
  ({ vertical }) => css`
    ${tw`
      flex
      flex-wrap
      justify-center
      items-start
      mb-6
      all-child:(mx-2 my-3)
    `}
    ${vertical && tw`flex-grow flex-col items-stretch all-child:(mx-auto w-full)`}

    & > & {
      margin: 0;
    }
  `
)

const FlexCard = tw(Card)`
  flex
  flex-col
  flex-grow
  items-center
  min-w-60
  max-w-lg
  px-6
  py-4
  space-y-2
`

const Field: FC<{ label: string; value: string }> = ({ label, value }) => (
  <div className="flex flex-row justify-center w-full space-x-2">
    <span className="flex-1 w-full text-right text-gray-500">{label}</span>
    <span className="flex-1 w-full">{value}</span>
  </div>
)

const Title = tw.div`text-2xl`
const GridCols1 = tw.ul`grid grid-flow-row grid-cols-1 md:grid-cols-1 grid-rows-1 gap-x-5 list-disc list-inside`
const GridCols3 = tw.ul`grid grid-flow-row grid-cols-2 md:grid-cols-3 grid-rows-1 gap-x-5 list-disc list-inside`

const PokemonDetailPage: NextPage<PokemonProps> = ({ pokemon }) => {
  if (!pokemon) {
    return <Title data-testid={TEST_ID_NO_FOUND}>No found pokemon</Title>
  }

  const {
    id,
    name,
    sprites,
    types,
    height,
    weight,
    abilities,
    stats,
    moves,
    game_indices,
  } = pokemon

  return (
    <>
      <Head>
        <title>{`${name} | Pokemon`}</title>
      </Head>
      <div className="flex justify-center space-x-4 mb-6">
        <div className="flex-none bg-white shadow p-4">
          <Image src={sprites.front_default} width={115} height={115} />
        </div>
        <div className="flex-row space-y-3">
          <div>
            <h2 className="text-2xl font-bold text-red-400">#{id}</h2>
            <h1 className="text-3xl capitalize" data-testid={TEST_ID_NAME_POKEMON}>
              {name}
            </h1>
          </div>

          <div className="flex-none flex-wrap">
            {types.map(({ type: { name } }, index) => (
              <TypeChip key={index} type={name} />
            ))}
          </div>
        </div>
      </div>

      <CardsContainer>
        <CardsContainer vertical>
          <FlexCard>
            <Title>Size</Title>

            <Field label="Height" value={`${getUnit(height)} m`} />

            <Field label="Weight" value={`${getUnit(weight)} kg`} />
          </FlexCard>

          <FlexCard>
            <Title>Abilities</Title>

            <GridCols1>
              {abilities.map(({ ability: { name } }, index) => (
                <li key={index}>{name}</li>
              ))}
            </GridCols1>
          </FlexCard>
        </CardsContainer>

        <FlexCard>
          <Title>Types defenses</Title>

          <DamageEffectiveness types={types} />
        </FlexCard>

        <FlexCard css={tw`lg:min-w-80`}>
          <Title>Stats</Title>

          {stats.map(({ stat: { name }, base_stat }, index) => {
            const { percentage } = calculateStatPoints(name, base_stat)
            return (
              <div key={index} className="w-full">
                <span className="capitalize">{name}</span>

                <ProgressBar
                  key={index}
                  size="sm"
                  percentage={percentage}
                  backgroundColor={getColorBaseStat(base_stat)}
                >
                  {base_stat}
                </ProgressBar>
              </div>
            )
          })}
        </FlexCard>

        <FlexCard>
          <Title>Games where appears</Title>

          <GridCols3>
            {game_indices.map(({ version: { name } }, index) => (
              <li key={index}>{name}</li>
            ))}
          </GridCols3>
        </FlexCard>

        <FlexCard>
          <Title>Moves</Title>

          <GridCols3>
            {moves.map(({ move: { name } }, index) => (
              <li key={index}>{name}</li>
            ))}
          </GridCols3>
        </FlexCard>
      </CardsContainer>
    </>
  )
}

export const getServerSideProps: GetServerSideProps<PokemonProps> = async ({
  params: { idOrName },
}) => {
  const pokemon = await getPokemon(idOrName as string).catch(() => undefined)

  if (!pokemon) {
    return {
      props: {
        error: {
          statusCode: 404,
          message: 'Not found',
        },
      },
    }
  }

  return {
    props: { pokemon },
  }
}

export default PokemonDetailPage
