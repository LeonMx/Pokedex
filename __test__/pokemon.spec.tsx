import { render, cleanup, screen } from '@testing-library/react'
import PokemonDetail, { TEST_ID_NAME_POKEMON, TEST_ID_NO_FOUND } from 'pages/pokemon/[idOrName]'
import { getPokemon } from 'api/pokemon'
import { Pokemon } from 'utils/types'
import { mockConsoleMethod } from 'utils/utils'

afterEach(cleanup)

console.warn = jest.fn(mockConsoleMethod(console.warn))
console.error = jest.fn(mockConsoleMethod(console.error))

const resolvePokemon = (idOrName: string | number): Promise<Pokemon> =>
  getPokemon(idOrName).catch(() => undefined)

describe('Pokemon', () => {
  it('Found pokemon', async () => {
    const pokemon = await resolvePokemon('1')
    const { asFragment } = render(<PokemonDetail pokemon={pokemon} />)
    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByTestId(TEST_ID_NAME_POKEMON)).toHaveTextContent(pokemon.name)
  })

  it('No found pokemon', async () => {
    const pokemon = await resolvePokemon('test')
    render(<PokemonDetail pokemon={pokemon} />)
    expect(screen.getAllByTestId(TEST_ID_NO_FOUND)).toHaveLength(1)
  })
})
