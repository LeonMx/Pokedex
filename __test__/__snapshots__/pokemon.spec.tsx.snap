// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Found pokemon 1`] = `
<DocumentFragment>
  <div
    class="idOrName___StyledDiv2-sc-5sbjqh-9 cZnTYL"
  >
    <div
      class="idOrName___StyledDiv3-sc-5sbjqh-10 bhtdmV"
    >
      <div
        style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
      >
        <div
          style="box-sizing: border-box; display: block; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            role="presentation"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTE1IiBoZWlnaHQ9IjExNSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
            style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
          />
        </div>
        <img
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
      </div>
    </div>
    <div
      class="idOrName___StyledDiv4-sc-5sbjqh-11 jZmcZS"
    >
      <div>
        <h2
          class="idOrName___StyledH-sc-5sbjqh-12 gWcCDd"
        >
          #1
        </h2>
        <h1
          class="idOrName___StyledH2-sc-5sbjqh-13 RnExT"
          data-testid="pokemon_name"
        >
          bulbasaur
        </h1>
      </div>
      <div
        class="idOrName___StyledDiv5-sc-5sbjqh-14 gswNHl"
      >
        <span
          class="Badge___StyledSpan-ok0bc8-0 cqoDdZ TypeChip___StyledBadge-sc-1if4k9o-0 kwrHCk"
        >
          grass
        </span>
        <span
          class="Badge___StyledSpan-ok0bc8-0 gqjKge TypeChip___StyledBadge-sc-1if4k9o-0 kwrHCk"
        >
          poison
        </span>
      </div>
    </div>
  </div>
  <div
    class="idOrName__CardsContainer-sc-5sbjqh-0 jPJrMR"
  >
    <div
      class="idOrName__CardsContainer-sc-5sbjqh-0 hraCyU"
    >
      <div
        class="Card___StyledDiv-sc-15iomcm-0 dpmhls idOrName__FlexCard-sc-5sbjqh-1 fuGMUl"
      >
        <div
          class="idOrName__Title-sc-5sbjqh-6 bywpKM"
        >
          Size
        </div>
        <div
          class="idOrName___StyledDiv-sc-5sbjqh-3 daXHih"
        >
          <span
            class="idOrName___StyledSpan-sc-5sbjqh-4 chDzbe"
          >
            Height
          </span>
          <span
            class="idOrName___StyledSpan2-sc-5sbjqh-5 hViBoq"
          >
            0.70 m
          </span>
        </div>
        <div
          class="idOrName___StyledDiv-sc-5sbjqh-3 daXHih"
        >
          <span
            class="idOrName___StyledSpan-sc-5sbjqh-4 chDzbe"
          >
            Weight
          </span>
          <span
            class="idOrName___StyledSpan2-sc-5sbjqh-5 hViBoq"
          >
            6.90 kg
          </span>
        </div>
      </div>
      <div
        class="Card___StyledDiv-sc-15iomcm-0 dpmhls idOrName__FlexCard-sc-5sbjqh-1 fuGMUl"
      >
        <div
          class="idOrName__Title-sc-5sbjqh-6 bywpKM"
        >
          Abilities
        </div>
        <ul
          class="idOrName__GridCols1-sc-5sbjqh-7 kqEYmf"
        >
          <li>
            overgrow
          </li>
          <li>
            chlorophyll
          </li>
        </ul>
      </div>
    </div>
    <div
      class="Card___StyledDiv-sc-15iomcm-0 dpmhls idOrName__FlexCard-sc-5sbjqh-1 fuGMUl"
    >
      <div
        class="idOrName__Title-sc-5sbjqh-6 bywpKM"
      >
        Types defenses
      </div>
      <svg
        class="Spinner___StyledSvg-sc-1tdtnot-0 jyURzq"
        fill="none"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          class="Spinner___StyledCircle-sc-1tdtnot-1 iXLzwL"
          cx="12"
          cy="12"
          r="10"
          stroke="currentColor"
          stroke-width="4"
        />
        <path
          class="Spinner___StyledPath-sc-1tdtnot-2 bQXkHa"
          d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="Card___StyledDiv-sc-15iomcm-0 dpmhls idOrName__FlexCard-sc-5sbjqh-1 idOrName___StyledFlexCard-sc-5sbjqh-2 fuGMUl kfFPHR"
    >
      <div
        class="idOrName__Title-sc-5sbjqh-6 bywpKM"
      >
        Stats
      </div>
      <div
        class="idOrName___StyledDiv6-sc-5sbjqh-15 zHTwa"
      >
        <span
          class="idOrName___StyledSpan3-sc-5sbjqh-16 kaFMpx"
        >
          hp
        </span>
        <div
          class="ProgressBar__StyledContainer-sc-1tzcbb6-0 htWqZh"
        >
          <div
            class="ProgressBar__StyledFiller-sc-1tzcbb6-1 fmObxI"
            color="#fff"
          >
            45
          </div>
        </div>
      </div>
      <div
        class="idOrName___StyledDiv6-sc-5sbjqh-15 zHTwa"
      >
        <span
          class="idOrName___StyledSpan3-sc-5sbjqh-16 kaFMpx"
        >
          attack
        </span>
        <div
          class="ProgressBar__StyledContainer-sc-1tzcbb6-0 htWqZh"
        >
          <div
            class="ProgressBar__StyledFiller-sc-1tzcbb6-1 jdwXHo"
            color="#fff"
          >
            49
          </div>
        </div>
      </div>
      <div
        class="idOrName___StyledDiv6-sc-5sbjqh-15 zHTwa"
      >
        <span
          class="idOrName___StyledSpan3-sc-5sbjqh-16 kaFMpx"
        >
          defense
        </span>
        <div
          class="ProgressBar__StyledContainer-sc-1tzcbb6-0 htWqZh"
        >
          <div
            class="ProgressBar__StyledFiller-sc-1tzcbb6-1 jdwXHo"
            color="#fff"
          >
            49
          </div>
        </div>
      </div>
      <div
        class="idOrName___StyledDiv6-sc-5sbjqh-15 zHTwa"
      >
        <span
          class="idOrName___StyledSpan3-sc-5sbjqh-16 kaFMpx"
        >
          special-attack
        </span>
        <div
          class="ProgressBar__StyledContainer-sc-1tzcbb6-0 htWqZh"
        >
          <div
            class="ProgressBar__StyledFiller-sc-1tzcbb6-1 hbjpwR"
            color="#fff"
          >
            65
          </div>
        </div>
      </div>
      <div
        class="idOrName___StyledDiv6-sc-5sbjqh-15 zHTwa"
      >
        <span
          class="idOrName___StyledSpan3-sc-5sbjqh-16 kaFMpx"
        >
          special-defense
        </span>
        <div
          class="ProgressBar__StyledContainer-sc-1tzcbb6-0 htWqZh"
        >
          <div
            class="ProgressBar__StyledFiller-sc-1tzcbb6-1 hbjpwR"
            color="#fff"
          >
            65
          </div>
        </div>
      </div>
      <div
        class="idOrName___StyledDiv6-sc-5sbjqh-15 zHTwa"
      >
        <span
          class="idOrName___StyledSpan3-sc-5sbjqh-16 kaFMpx"
        >
          speed
        </span>
        <div
          class="ProgressBar__StyledContainer-sc-1tzcbb6-0 htWqZh"
        >
          <div
            class="ProgressBar__StyledFiller-sc-1tzcbb6-1 fmObxI"
            color="#fff"
          >
            45
          </div>
        </div>
      </div>
    </div>
    <div
      class="Card___StyledDiv-sc-15iomcm-0 dpmhls idOrName__FlexCard-sc-5sbjqh-1 fuGMUl"
    >
      <div
        class="idOrName__Title-sc-5sbjqh-6 bywpKM"
      >
        Games where appears
      </div>
      <ul
        class="idOrName__GridCols3-sc-5sbjqh-8 jZLtT"
      >
        <li>
          red
        </li>
        <li>
          blue
        </li>
        <li>
          yellow
        </li>
        <li>
          gold
        </li>
        <li>
          silver
        </li>
        <li>
          crystal
        </li>
        <li>
          ruby
        </li>
        <li>
          sapphire
        </li>
        <li>
          emerald
        </li>
        <li>
          firered
        </li>
        <li>
          leafgreen
        </li>
        <li>
          diamond
        </li>
        <li>
          pearl
        </li>
        <li>
          platinum
        </li>
        <li>
          heartgold
        </li>
        <li>
          soulsilver
        </li>
        <li>
          black
        </li>
        <li>
          white
        </li>
        <li>
          black-2
        </li>
        <li>
          white-2
        </li>
      </ul>
    </div>
    <div
      class="Card___StyledDiv-sc-15iomcm-0 dpmhls idOrName__FlexCard-sc-5sbjqh-1 fuGMUl"
    >
      <div
        class="idOrName__Title-sc-5sbjqh-6 bywpKM"
      >
        Moves
      </div>
      <ul
        class="idOrName__GridCols3-sc-5sbjqh-8 jZLtT"
      >
        <li>
          razor-wind
        </li>
        <li>
          swords-dance
        </li>
        <li>
          cut
        </li>
        <li>
          bind
        </li>
        <li>
          vine-whip
        </li>
        <li>
          headbutt
        </li>
        <li>
          tackle
        </li>
        <li>
          body-slam
        </li>
        <li>
          take-down
        </li>
        <li>
          double-edge
        </li>
        <li>
          growl
        </li>
        <li>
          strength
        </li>
        <li>
          mega-drain
        </li>
        <li>
          leech-seed
        </li>
        <li>
          growth
        </li>
        <li>
          razor-leaf
        </li>
        <li>
          solar-beam
        </li>
        <li>
          poison-powder
        </li>
        <li>
          sleep-powder
        </li>
        <li>
          petal-dance
        </li>
        <li>
          string-shot
        </li>
        <li>
          toxic
        </li>
        <li>
          rage
        </li>
        <li>
          mimic
        </li>
        <li>
          double-team
        </li>
        <li>
          defense-curl
        </li>
        <li>
          light-screen
        </li>
        <li>
          reflect
        </li>
        <li>
          bide
        </li>
        <li>
          sludge
        </li>
        <li>
          skull-bash
        </li>
        <li>
          amnesia
        </li>
        <li>
          flash
        </li>
        <li>
          rest
        </li>
        <li>
          substitute
        </li>
        <li>
          snore
        </li>
        <li>
          curse
        </li>
        <li>
          protect
        </li>
        <li>
          sludge-bomb
        </li>
        <li>
          mud-slap
        </li>
        <li>
          giga-drain
        </li>
        <li>
          endure
        </li>
        <li>
          charm
        </li>
        <li>
          swagger
        </li>
        <li>
          fury-cutter
        </li>
        <li>
          attract
        </li>
        <li>
          sleep-talk
        </li>
        <li>
          return
        </li>
        <li>
          frustration
        </li>
        <li>
          safeguard
        </li>
        <li>
          sweet-scent
        </li>
        <li>
          synthesis
        </li>
        <li>
          hidden-power
        </li>
        <li>
          sunny-day
        </li>
        <li>
          rock-smash
        </li>
        <li>
          facade
        </li>
        <li>
          nature-power
        </li>
        <li>
          ingrain
        </li>
        <li>
          knock-off
        </li>
        <li>
          secret-power
        </li>
        <li>
          grass-whistle
        </li>
        <li>
          bullet-seed
        </li>
        <li>
          magical-leaf
        </li>
        <li>
          natural-gift
        </li>
        <li>
          worry-seed
        </li>
        <li>
          seed-bomb
        </li>
        <li>
          energy-ball
        </li>
        <li>
          leaf-storm
        </li>
        <li>
          power-whip
        </li>
        <li>
          captivate
        </li>
        <li>
          grass-knot
        </li>
        <li>
          venoshock
        </li>
        <li>
          round
        </li>
        <li>
          echoed-voice
        </li>
        <li>
          grass-pledge
        </li>
        <li>
          work-up
        </li>
        <li>
          grassy-terrain
        </li>
        <li>
          confide
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
